### PR TITLE
<for the master branch> tests: (lsfs) skip some test cases if making a NETLINK_SOCK_DIAG socket and/or sending a message via the socket are not allowed

### PR DESCRIPTION
**Pull request by masatake**
Do the same as #2841 but for the master branch.

---

Close https://github.com/util-linux/util-linux/issues/2822

As reported on https://github.com/util-linux/util-linux/issues/2822, some test cases that require a NETLINK_SOCK_DIAG socket for the AF_UNIX family failed. The platform where the test ran might disallow the test cases from making a NETLINK_SOCK_DIAG socket and/or sending a message via it.

This pull request enhances test_mkfds to make a NETLINK_SOCK_DIAG socket for the AF_UNIX family and send a message via the socket. If running test_mkfds fails, the test cases are skipped.

### CODE REVIEW

File: tests/helpers/test_mkfds.c [hunk 1]
```diff
@@ -3213,6 +3216,99 @@ static void free_mmap(const struct factory * factory _U_, void *data)
...
+		udr = (struct unix_diag_req) {
+			.sdiag_family = AF_UNIX,
+			.udiag_states = -1, /* set the all bits. */
+			.udiag_show = UDIAG_SHOW_NAME | UDIAG_SHOW_PEER | UNIX_DIAG_SHUTDOWN,
+		};
+		req = &udr;
+		reqlen = sizeof(udr);
+	}
+
+	int e = send_diag_request(diagsd, req, reqlen);
```
Comment (karelzak) [inline tests/helpers/test_mkfds.c:3281]: This is code for the test, so I don't think we have to be pedantic but note that variables should be defined at the beginning of the code block.

File: tests/helpers/test_mkfds.c [hunk 2]
```diff
@@ -3213,6 +3216,99 @@
...
+			err(EXIT_EACCES, "failed in sendmsg()");
+		if (errno == ENOENT)
+			err(EXIT_ENOENT, "failed in sendmsg()");
+		err(EXIT_FAILURE, "failed in sendmsg()");
+	}
+
+
+	if (diagsd != fdescs[0].fd) {
+		if (dup2(diagsd, fdescs[0].fd) < 0) {
+			int e = errno;
```
[the remark this answers is not in the document]
Comment (karelzak) [reply tests/helpers/test_mkfds.c:3295]: This is true and not readable and elegant to use `e` in this way.

**Comment by karelzak** [comment]
It seems ASAN does not like it:
```
==102406==ERROR: AddressSanitizer: stack-use-after-scope on address 0x7fef365000e0 at pc 0x7fef39265f6e bp 0x7ffc05444980 sp 0x7ffc05444128
READ of size 24 at 0x7fef365000e0 thread T0
    #0 0x7fef39265f6d in read_iovec ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:1082
    #1 0x7fef392be995 in read_msghdr ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:3307
    #2 0x7fef392c03d5 in __interceptor_sendmsg ../../../../src/libsanitizer/sanitizer_common/sanitizer_common_interceptors.inc:3324
    #3 0x5596c543f8f8 in send_diag_request tests/helpers/test_mkfds.c:3243
    #4 0x5596c543f8f8 in make_sockdiag tests/helpers/test_mkfds.c:3281
    #5 0x5596c545c316 in main tests/helpers/test_mkfds.c:4517
    #6 0x7fef38629d8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: c289da5071a3399de893d2af81d6a30c62646e1e)
    #7 0x7fef38629e3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: c289da5071a3399de893d2af81d6a30c62646e1e)
    #8 0x5596c5439004 in _start (/home/runner/work/util-linux/util-linux/test_mkfds+0x37004) (BuildId: c29fbb781283139cd1fc43be603a6265c4781847)
```

**Comment by masatake** [comment]
I'll revise this pull request based on #2888.

**Comment by masatake** [comment]
Updated.
Now this pull request contains the merged changes (#2841 and #2888) for the stable branch.

**Pull request merged**